module/modmanager/manager_test.go

```diff
@@ -1322,3 +1337,52 @@ func TestBadModuleFailsFast(t *testing.T) {
 
 	test.That(t, err.Error(), test.ShouldContainSubstring, "module test-module exited too quickly after attempted startup")
 }
+
+func TestModularDiscovery(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+
+	modPath := rtestutils.BuildTempModule(t, "module/testmodule")
+
+	modCfg := config.Module{
+		Name:    "test-module",
+		ExePath: modPath,
+	}
+
+	parentAddr := setupSocketWithRobot(t)
+
+	mgr := setupModManager(t, ctx, parentAddr, logger, modmanageroptions.Options{UntrustedEnv: false})
+
+	err := mgr.Add(ctx, modCfg)
+	test.That(t, err, test.ShouldBeNil)
+
+	// The helper model implements actual (foobar) discovery
+	reg, ok := resource.LookupRegistration(generic.API, resource.NewModel("rdk", "test", "helper"))
+	test.That(t, ok, test.ShouldBeTrue)
+	test.That(t, reg, test.ShouldNotBeNil)
+
+	// Check that the Discover function is registered and make call
+	test.That(t, reg.Discover, test.ShouldNotBeNil)
+	result, err := reg.Discover(ctx, logger)
+	test.That(t, err, test.ShouldBeNil)
+	t.Log("Discovery result: ", result)
+
+	// Format result
+	jsonData, err := json.Marshal(result)
+	test.That(t, err, test.ShouldBeNil)
+	// Debug: print the JSON data
+	t.Logf("Raw JSON: %s", string(jsonData))
+
+	var discoveryResult TestDiscoveryResult
```

Comment:
per my suggestion above

```suggestion
	var discoveryResult testDiscoveryResult
```